components/ai_chat/core/browser/engine/engine_consumer_conversation_api.cc

```diff
@@ -215,37 +236,128 @@ void EngineConsumerConversationAPI::GenerateAssistantResponse(
+            bool is_large = false;
+            size_t content_size = 0;
+            for (const auto& content : tool_event->output.value()) {
+              if (content->is_image_content_block()) {
+                is_large = true;
+                break;
+              } else if (content->is_text_content_block()) {
+                content_size += content->get_text_content_block()->text.size();
+                if (content_size >= 1000) {
+                  is_large = true;
+                  break;
+                }
+              }
+            }
+            if (large_event_count < kMaxCountLargeToolUseEvents || !is_large) {
+              if (is_large) {
+                large_event_count++;
+              }
+              std::vector<mojom::ContentBlockPtr> tool_result_content;
+              for (const auto& item : tool_event->output.value()) {
+                tool_result_content.push_back(item.Clone());
+              }
+              tool_result.content = std::move(tool_result_content);
+            } else {
+              tool_result.content = std::vector<std::string>{
+                  "[Large result removed to save space for subsequent "
+                  "results]"};
+            }
+          }
+          tool_results.emplace_back(std::move(tool_result));
+        }
+      }
+    }
 
     // TODO(petemill): Shouldn't the server handle the map of mojom::ActionType
     // to prompts in addition to SUMMARIZE_PAGE (e.g. PARAPHRASE, EXPLAIN,
     // IMPROVE, etc.)
     if (message->action_type == mojom::ActionType::SUMMARIZE_PAGE) {
       event.type = ConversationEventType::RequestSummary;
-      event.content = {""};
+      event.content = std::vector<std::string>{""};
     } else {
       event.type = ConversationEventType::ChatMessage;
     }
-    conversation.push_back(std::move(event));
+
+    // Tool result should be sorted after the message
+    if (!tool_results.empty()) {
+      for (auto& tool_event : tool_results) {
+        conversation_message_insertion_it = conversation.insert(
+            conversation_message_insertion_it, std::move(tool_event));
+      }
+    }
+
+    conversation_message_insertion_it = conversation.insert(
+        conversation_message_insertion_it, std::move(event));
+
+    if (message->selected_text.has_value() &&
+        !message->selected_text->empty()) {
+      conversation_message_insertion_it = conversation.insert(
+          conversation_message_insertion_it,
+          ConversationEvent(
+              ConversationEventRole::User, ConversationEventType::PageExcerpt,
+              std::vector<std::string>{message->selected_text.value()}));
+    }
+
+    // Insert any events before the message
+    if (!events_before_message.empty()) {
+      conversation_message_insertion_it = conversation.insert(
+          conversation_message_insertion_it,
+          std::make_move_iterator(events_before_message.begin()),
+          std::make_move_iterator(events_before_message.end()));
+    }
```

Comment:
Hey, so the purpose of this seems to be that these events end up in the reverse order to what we've got here. Rather than doing this double reverse, could we just `push_back` but in the opposite order to what we have here?

It took me quite a while to understand what this code was trying to do

Reply:
The purpose of the reverse iteration is to remove large tool output at the beginning of the conversation in favor of large tool output at the end. There is a comment (quite far) above:

```
  // We're going to iterate over the conversation entries and
  // build a list of events for the remote API.
  // The iteration will be in reverse order so we can prioritize data
  // in the most recent messages.
  // Whilst iterating messages in reverse, we'll need to insert events before
  // the previous message, so we'll keep track of this position.
```

Reply:
Thanks for doing that - I find the new approach much easier to follow. I left a bunch of comments on the commit :smile: